ruby/lib/gherkin/query.rb

```diff
@@ -20,34 +27,42 @@ def update_feature(feature)
       store_nodes_location(feature.tags)
 
       feature.children.each do |child|
-        update_rule(child.rule) if child.rule
-        update_background(child.background) if child.background
-        update_scenario(child.scenario) if child.scenario
+        update_rule(feature, child.rule) if child.rule
+        update_background(feature, child.background) if child.background
+        update_scenario(feature, child.rule, child.scenario) if child.scenario
       end
     end
 
-    def update_rule(rule)
+    def update_rule(feature, rule)
       return if rule.nil?
       store_nodes_location(rule.tags)
-
       rule.children.each do |child|
-        update_background(child.background) if child.background
-        update_scenario(child.scenario) if child.scenario
+        update_background(rule, child.background) if child.background
+        update_scenario(feature, rule, child.scenario) if child.scenario
       end
     end
 
-    def update_background(background)
+    def update_background(parent, background)
       update_steps(background.steps)
+      parent_id = parent.respond_to?(:id) ? parent.id : parent.object_id # Cucumber::Messages::Feature doesn't have an id :(
+      @background_scopes[parent_id] = background
     end
 
-    def update_scenario(scenario)
+    def update_scenario(feature, rule, scenario)
       store_node_location(scenario)
       store_nodes_location(scenario.tags)
       update_steps(scenario.steps)
       scenario.examples.each do |examples|
         store_nodes_location(examples.tags || [])
         store_nodes_location(examples.table_body || [])
       end
+
+      @scenario_parent_locations[scenario.id] = [
```

Comment:
To save time, as this is the only bit of the new API that's completely new, can you explain it in a line or two of prose so I don't need to dig into the whole API surface for gherkin

Reply:
Sure, this basically is the meat of the entire effort to fix the long-standing Cucumber 4.0 regression https://github.com/cucumber/cucumber-ruby/issues/1469

This block of code is collating all the line numbers of all the parents of the given scenario:
1. The Feature (and its optional description lines)
2. The optional Background and its steps for the _Scenario_
3. The optional Rule line
4. The optional Background and its steps for the _Rule_

It stores these line numbers data structure -- keyed to the scenario -- for later lookup.

Reply:
Could you dump an example structure. I'm anticipating something like (With dodgy linespacing delib)

```
1 - Feature: Foo
2 - 
3 - Desc of feature
4 - 
5 - Background:
6 -  Given background one
7 - Scenario: Bar
8 - Given baz
9 - Then foo
```

Produces

```
{ feature: [1,3], background: [5], Rule: [], RuleBackground: []}
```

Reply:
And for my benefit. We're only looking at **parents** here because `locations` returns everything from the scenario down right?

Reply:
@luke-hill Sure, I just changed the `expect` to `puts(...).inspect` in `spec/gherkin/query_spec.rb#L101`, which tests the method with the "Scenario: a ruled scenario" scenario and this is the result (I added line numbers here for easier output interpretting):

```gherkin
1
2    @feature-tag
3    Feature: my feature
4
5      Background:
6        Given a passed background step
7
8      @scenario-tag
9      Scenario: my scenario
10        Given a passed step
11
12      Scenario Outline: with examples
13        Given a <Status> step
14
15        @examples-tag
16        Examples:
17          | Status |
18          | passed |
19          | failed |
20
21      @rule-tag
22      Rule: this is a rule
23        Background:
24          Given the passed step in the rule background
25
26        @ruled-scenario-tag
27        Scenario: a ruled scenario
28          Given a step in the ruled scenario
```

```ruby
[#<Cucumber::Messages::Location:0x00007f4b2805a620 @line=3, @column=5>,
#<Cucumber::Messages::Location:0x00007f4b28158ea0 @line=5, @column=7>,
#<Cucumber::Messages::Location:0x00007f4b2805bd68 @line=22, @column=7>,
#<Cucumber::Messages::Location:0x00007f4b2804efc8 @line=23, @column=9>]
```

To answer your second question, you might notice that some lines are missing, like tag lines (2, 21, 26) and step lines (6, 24, 28). These lines are already matched correctly in `Cucumber::Core::Compiler#source_lines_for_pickle`. The companion PR over there adds a `#parent_source_lines_for_pickle` that uses this PR's `Gherkin::Query#scenario_parent_locations` method to match the last remaining missing pieces, the Feature, the Background, the Rule, and the Rule Background` lines, which is what you see in the output here.

I first looked at extending `#source_lines_for_pickle` to return everything, but other parts of the cucumber internals were relying on only the source lines being returned, not these extra parent lines, so we needed to keep them separate in the `C::C::TestCase` data structure. Hence, this additional method.

Reply:
Absolutely perfectly explained.

I personally dislike pos array returning bits over a hash. But I'm 99.9% certain you're going to say the other methods work the same way. i.e. it's magic that the first item is the feature ref

Reply:
I agree with you on that one, but the good news is that we don't actually care about where the lines come from, only that they match the given scenario. I.e. we could append a #shuffle to the array before returning it, and it wouldn't affect a thing.